### ISSUE TITLE
Wrong bounds for floating-point variables

### ISSUE THREAD

**Issue by krzku** [MiniZinc/libminizinc#316]

**Comment by krzku** [comment]
I have recently tried the attached problem and both Gecode and JaCoP reported that the problem is unsatisfied while the problem has a solution. Encoding the problem directly in Java for JaCoP provides correct solution. My guess is that the bounds for floating-point variables are not computed correctly. In interval arithmetic one needs to take care of proper rounding of every intermediate result in order to preserve the property that the final results are guaranteed. The appropriate rounding is called outward rounding, that is, the lower bound value must be always rounded down (towards -inf) and the upper bound value has to be rounded up (towards +inf). This can be implemented in hardware by selecting correct rounding mode for floating-point operations or in software. It **seems** that the compiler does not follow this rule.

/Kris

[birthday_paradox.mzn.zip](https://github.com/MiniZinc/libminizinc/files/3619929/birthday_paradox.mzn.zip)

**Comment by guidotack** [comment]
MiniZinc doesn't guarantee correct rounding in that sense (the type is called float for a reason, it's not interval or real variables but all the arithmetic is simply done as floating point). This is a limitation at the moment but consistent with what most MIP/LP solvers do. We may add a "real" type in the future that would mitigate this problem. Note that part of the problem may simply be the round-trip from decimal (in the model) to binary (internal) to decimal (in the FlatZinc) to binary (in the solver), which can cause rounding issues.

**Comment by krzku** [comment]
Thanks Guido for the explanation. Of course the problem can be also caused by the round-trip from decimal to binary and this is why I wrote "seems" in bold. I did not investigated this much but I tried to round down and round up intervals for variables in the solver and it improved result for JaCop a little bit. It could solve the problem for 35 persons instead of 34 ;). That is a small improvement.

I think the name float of course suggest the discrete representation of reals but it does not explains wrong rounding. 

Yes, it is fine with MIP/LP solvers since they do not use interval arithmetic.

Best,
/Kris
